plutus-contract/test/Spec/MustSpendAtLeast.hs

```diff
@@ -0,0 +1,116 @@
+                             defaultCheckOptions, w1, (.&&.))
+import Plutus.Trace qualified as Trace
+import Plutus.V1.Ledger.Api (Datum (Datum), ScriptContext, Validator, ValidatorHash)
+import Plutus.V1.Ledger.Scripts (unitRedeemer)
+import PlutusTx qualified
+import PlutusTx.Prelude as P (traceError)
+import Prelude hiding (not)
+
+tests :: TestTree
+tests =
+    testGroup "MustSpendAtLeast"
+        [ entireScriptBalance
+        , lessThanScriptBalance
+        , higherThanScriptBalance
+        ]
+
+scriptBalance :: Integer
+scriptBalance = 25_000_000
+
+contract :: Integer -> Contract () Empty ContractError ()
+contract amt = do
+    let lookups1 = Constraints.plutusV1TypedValidatorLookups $ typedValidator amt
+        tx1 = Constraints.mustPayToTheScript () (Ada.lovelaceValueOf scriptBalance)
+    ledgerTx1 <- submitTxConstraintsWith lookups1 tx1
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx1
+    utxos <- utxosAt $ scrAddress amt
+    let orefs = fst <$> Map.toList utxos
+        lookups2 =
+            Constraints.plutusV1OtherScript (validatorScript amt)
+            <> Constraints.unspentOutputs utxos
+        tx2 =
+            foldMap (\oref -> Constraints.mustSpendScriptOutput oref unitRedeemer) orefs
+            <> Constraints.mustIncludeDatum (Datum $ PlutusTx.toBuiltinData amt)
+            <> Constraints.mustSpendAtLeast (Ada.lovelaceValueOf amt)
+    ledgerTx2 <- submitTxConstraintsWith @Void lookups2 tx2
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx2
+
+trace :: Integer -> Trace.EmulatorTrace ()
+trace amt = do
+    void $ Trace.activateContractWallet w1 $ contract amt
+    void $ Trace.waitNSlots 1
+
+entireScriptBalance :: TestTree
+entireScriptBalance = checkPredicateOptions
+    defaultCheckOptions
+    "Successful use of mustSpendAtLeast at script's exact balance"
+    (assertValidatedTransactionCount 2)
+    (void $ trace scriptBalance)
+
+lessThanScriptBalance :: TestTree
+lessThanScriptBalance =
+    let amt = scriptBalance - 1
+    in  checkPredicateOptions
+            defaultCheckOptions
+            "Successful use of mustSpendAtLeast below script's balance"
+            (assertValidatedTransactionCount 2)
+            (void $ trace amt)
+
+higherThanScriptBalance :: TestTree -- This should fail with a more accurate contract error than "OwnPubKeyMissing"
+higherThanScriptBalance =
+    let amt = scriptBalance + 1
+    in  checkPredicateOptions
+            defaultCheckOptions
+            "Fail validation when mustSpendAtLeast is greater than script's balance"
+            (assertContractError (contract amt) (Trace.walletInstanceTag w1) (\case { ConstraintResolutionContractError Constraints.OwnPubKeyMissing -> True; _ -> False }) "failed to throw error"
+            .&&. assertValidatedTransactionCount 1)
+            (void $ trace amt)
+
+{-# INLINEABLE mkValidator #-}
+mkValidator :: Integer -> () -> () -> ScriptContext -> Bool
```

Comment:
Did you actually meant `mkValidator :: Integer -> () -> ScriptContext -> Bool`?

Because when trying to unlock the script, you used `Constraints.mustIncludeDatum (Datum $ PlutusTx.toBuiltinData amt)`

Reply:
Changed to use datum